editor/src/clj/editor/fxui.clj

```diff
@@ -636,3 +663,274 @@
+          (.play anim)
+          (.consume e))))
+
+    KeyCode/ESCAPE
+    (when-not (= edit text)
+      (.consume e)
+      (swap-state dissoc :edit))
+
+    nil))
+
+(defn- handle-value-field-focused-changed [edit text swap-state on-value-changed to-value focused]
+  (when (and (not focused) (not= edit text))
+    (if-some [value (to-value edit)]
+      (do (swap-state #(-> % (assoc :value value) (dissoc :edit)))
+          (when on-value-changed (on-value-changed value)))
+      (swap-state dissoc :edit))))
+
+(def ^:private ext-with-value-field-text-props
+  (fx/make-ext-with-props
+    {:text (fx.prop/make
+             (fx.mutator/setter
+               (fn [^TextInputControl text-input text]
+                 (when-not (= text (.getText text-input))
+                   (.setText text-input text)
+                   (.selectAll text-input))))
+             fx.lifecycle/scalar)}))
+
+(defn- value-field-impl-final-step [{:keys [state swap-state text on-value-changed to-value component commit-on-enter]
+                                     :or {to-value identity}
+                                     :as props}]
+  (let [edit (:edit state text)]
+    {:fx/type ext-with-value-field-text-props
+     :props {:text edit}
+     :desc (-> props
+               (assoc :fx/type component
+                      :on-text-changed (fn/partial swap-state assoc :edit)
+                      :on-key-pressed (fn/partial handle-value-field-key-pressed edit text swap-state on-value-changed to-value commit-on-enter)
+                      :on-focused-changed (fn/partial handle-value-field-focused-changed edit text swap-state on-value-changed to-value))
+               (dissoc :state :swap-state :on-value-changed :to-value :text :component :commit-on-enter))}))
+
+(defn- stringify-value [f v]
+  (if (identical? v ::not-found)
+    ""
+    (str (f v))))
+
+(defn- value-field-impl-stringify-step [props]
+  {:fx/type ext-memo
+   :fn stringify-value
+   :args [(:to-string props str) (-> props :state :value)]
+   :key :text
+   :desc (-> props
+             (assoc :fx/type value-field-impl-final-step)
+             (dissoc :to-string :value))})
+
+(defn- make-value-field [component commit-on-enter props]
+  {:fx/type fx/ext-state
+   :initial-state {:value (:value props ::not-found)}
+   :desc (assoc props :fx/type value-field-impl-stringify-step
+                      :component component
+                      :commit-on-enter commit-on-enter)})
+
+(defn value-field
+  "Text field with value commit/reset semantics
+
+  Supports all :text-field props, plus:
+    :value               the edited value
+    :on-value-changed    value change callback
+    :to-string           value->string converter, default str
+    :to-value            string->value converter, default identity, returning
+                         nil implies value could not be converted"
+  [props]
+  (make-value-field text-field true props))
```

Comment:
`value-field` is a copy of the editor script implementation. I thought it would be possible to avoid code duplication here and reuse this value-view in the editor scripts, but unfortunately, there are some differences in the implementations that would take more time to resolve. The editor scripts' implementation uses different handling of focus change because it also needs to access the dialog window when the focus changes. It passes it to its value change handler so that in case the lua value change listener decides to open a new dialog, the current one will be set as its owner, transparently for the user. We probably want this here as well... eventually... But I thought this would be too out of scope for the keyboard shortcuts feature.